商品出品機能

shiki0501 opened the pull request:
# What
出品機能実装
# Why
出品機能を実装することでフリマアプリ内に出品の情報を保存し売買の基本情報を準備、取得するため

- ログイン状態のユーザーは、商品出品ページへ遷移できる動画
https://gyazo.com/a79d526e04449651b28e9672228e0478
- 金額が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/dd460750c185100a491e9ad03f912cc7
- 正しく情報を記入すると、商品を出品できる動画
https://gyazo.com/3ca2429a2a28f318c87d86d86a1cf9cb
- 入力に問題のある状態では商品は出品できず、エラーメッセージが出力される動画
https://gyazo.com/c789b956f5ab721b5e2c14491de586ab
- ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/caa46990719393b928ab899917522c38
- テスト結果の画像
https://gyazo.com/851d80fbdf188ebc5c92bb79452d8095

コードレビューのほどよろしくお願いします。

shiki0501 committed:
商品出品機能（出品遷移(ユーザー遷移管理未実装)、ActiveHash(選択できる状態)

shiki0501 committed:
Active Storangeの導入

shiki0501 committed:
（修正）ActiveHashの導入　この記述、今は必要なし

shiki0501 committed:
出品の遷移　未ログイン時のログイン画面遷移

shiki0501 committed:
販売手数料や販売利益の計算

shiki0501 committed:
出品データのデータベース保存機能

shiki0501 committed:
単体テストコードの実装,（rubocopにてコードの整理）

shiki0501 committed:
不要なコード削除、paramsへ:image追加（バリテーションのため）

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,30 @@
 class ItemsController < ApplicationController
   def index
+    @item = Item.order('created_at DESC')
```
```rb
@item = Item.order('created_at DESC')
```
こちらの記述は商品一覧機能の実装で記述する部分なので、一旦コメントアウトしておきましょう。
GitHubのレビューでは差分しか表示がされない仕様になっています。
そのため、商品一覧表示機能のレビューを依頼したときに変更した記述を見ることができない可能性があるためです。

TakashiKozuma commented on app/models/item.rb:
```diff
@@ -0,0 +1,41 @@
+  belongs_to :shipping_day
+  has_one_attached :image
+
+  with_options numericality: { other_than: 1 } do
+    validates :category_id
+    validates :status_id
+    validates :delivery_charge_id
+    validates :shipping_area_id
+    validates :shipping_day_id
+  end
+
+  with_options numericality: { other_than: 1, message: "can't be blank" } do
+    validates :category_id
+    validates :status_id
+    validates :delivery_charge_id
+    validates :shipping_area_id
+    validates :shipping_day_id
+  end
```
c```rb
  with_options numericality: { other_than: 1 } do
    validates :category_id
    validates :status_id
    validates :delivery_charge_id
    validates :shipping_area_id
    validates :shipping_day_id
  end

  with_options numericality: { other_than: 1, message: "can't be blank" } do
    validates :category_id
    validates :status_id
    validates :delivery_charge_id
    validates :shipping_area_id
    validates :shipping_day_id
  end
```
アクティブハッシュに関するバリデーションが重複しているので、削除しておき間そう。
具体的には「1以外では登録できない」というバリデーションが2回かかっています。

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,30 @@
 class ItemsController < ApplicationController
```
下記の実装条件を満たすように、コントローラーに記述を追加しましょう。
- ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移すること

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,30 @@
+
+  private
+
+  def user_params
+    params.require(:user)
+  end
```
```rb
  def user_params
    params.require(:user)
  end
```
このprivateメソッドは特に使用しないのであれば削除しておきましょう。

後に使用する場合でも、出品機能に関わらないものであれば、コメントアウトをしておきましょう。

TakashiKozuma commented on app/views/items/new.html.erb:
```diff
@@ -0,0 +1,159 @@
+    <h2 class="items-sell-title">商品の情報を入力</h2>
+    <%= form_with model: @item, local: true do |f| %>
+
+    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+    <%= render 'shared/error_messages', model: f.object %>
+    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```rb
    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
    <%= render 'shared/error_messages', model: f.object %>
    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
10,12行目のコメントアウトは実装案内なので削除しておきましょう。
実装案内は完成したアプリを人に見せる際に残しておくの望ましくないためです。
他にも「〜しましょう」のような実装案内があれば、実装時に削除するようにしていきましょう。

TakashiKozuma commented on config/routes.rb:
```diff
@@ -1,4 +1,7 @@
   devise_for :users
   root to: "items#index"
+  
+  resources :items 
```
```rb
resources :items
```
現状実装しているのはindex,new,createだけなので、onlyオプションを用いて、ルーティングを制限しましょう。
実装していない機能のルーティングを設定するのは、予期せぬエラーを起こすなどアプリの挙動に影響を与える場合があります。

TakashiKozuma commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,72 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price can't be blank")
+      end
+      it 'priceが¥300~¥9,999,999の範囲ないでないと登録できない' do
+        @item.price = '299'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not included in the list')
+      end
+      it 'priceが半角英数字でないと登録できない' do
+        @item.price = '５０００'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not included in the list')
+      end
```
```rb
      it 'priceが¥300~¥9,999,999の範囲ないでないと登録できない' do
        @item.price = '299'
        @item.valid?
        expect(@item.errors.full_messages).to include('Price is not included in the list')
      end
      it 'priceが半角英数字でないと登録できない' do
        @item.price = '５０００'
        @item.valid?
        expect(@item.errors.full_messages).to include('Price is not included in the list')
      end
```
priceは数値を扱うカラムであるため、``は外して`5000`などのように数値として扱えるように変更しましょう。

TakashiKozuma replied:
`半角数字でなければ保存できないこと`を確かめるため、以下のように3点をテストする必要があります。
不足しているテストを追加で実装しましょう。

・全角文字では登録できないこと
・半角英数混合では登録できないこと
・半角英語だけでは登録できないこと

TakashiKozuma commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,72 @@
+        @item.shipping_day_id = ''
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Shipping day can't be blank")
+      end
```
アクティブハッシュに関するカラムは、「1では登録できない」というバリデーション を付与しているため、そちらもテストを行いましょう。
5つのカラム全てに対してテストが必要です。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
コメントを致しましたので、ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

shiki0501 committed:
不要コード削除、devise ヘルパーメソッドによる未ログイン動作の確認、priceバリテーションの見直し

shiki0501 commented:
コードレビューありがとうございました。
修正いたしましたので、再度レビューのほどよろしくお願いいたします。

- ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/7d6a76d5dbcc99b29a2ee6387da77bd0
- テスト結果の画像
https://gyazo.com/4d5c0b7400734832d2f1f858d0540fb0

KenyaMatsumoto commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,26 @@
 class ItemsController < ApplicationController
+  before_action :authenticate_user!, only: [:new]
```
```ruby
   before_action :authenticate_user!, only: [:new]
```

セキュリティ面を考慮して、こちらの記述を```createアクション```にも適用させましょう！

createアクションにも適用する理由：
create、update、destroyアクションは、今回のFURIMAにおいて対応するビューファイルがなく（create.html.erbなど）、一見不正なアクセスができないように思える処理ですが、「検証ツールを使用する」「URLの直接入力をする」といった行為で不正なアクセスができてしまうことがあるためです。

KenyaMatsumoto commented on db/migrate/20210718014100_create_items.rb:
```diff
@@ -0,0 +1,18 @@
+      t.integer :shipping_area_id , null: false
+      t.integer :shipping_day_id ,null: false
+      t.integer :price, null: false
+      t.references :user
```
```ruby
 t.references :user
```
こちらのカラムのオプションが不足していることにより、user_idがなくても保存ができているかと存じます。
README.mdを見直し、マイグレーションファイルを訂正したあとに、rails db:rollback、マイグレーションの程よろしくお願いいたします。

KenyaMatsumoto commented on app/javascript/price.js:
```diff
@@ -0,0 +1,14 @@
+      PriceTax.innerHTML = Math.floor(price.value*Tax)
+      const PriceProfit = document.getElementById("profit")
+      PriceProfit.innerHTML = price.value-PriceTax.innerHTML
+    console.log(PriceTax)
```
```ruby
 console.log(PriceTax)
```
こちらのデバッグの記述は削除しましょう。

【理由】
開発時の挙動確認のための記述なので、実装後は必要ないためです。

KenyaMatsumoto commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,26 @@
+  def item_params
+    params.require(:item).permit(:image, :name, :description, :category_id, :status_id, :delivery_charge_id, :shipping_area_id,
+                                 :shipping_day_id, :price)
   end
```
```ruby
def item_params
    params.require(:item).permit(:image, :name, :description, :category_id, :status_id, :delivery_charge_id, :shipping_area_id,
                                 :shipping_day_id, :price)
  end
```
マイグレーションファイル修正後にmergeメソッドを使用して、user_idも保存しましょう！
現状はマイグレーションファイルの記述が異なっていますので、user_idがなくても商品の保存ができる仕様になっております。

KenyaMatsumoto commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,16 @@
+    shipping_area_id   { Faker::Number.between(from: 2, to: 3) }
+    shipping_day_id    { Faker::Number.between(from: 2, to: 3) }
+    price              { Faker::Number.between(from: 300, to: 9_999_999) }
+
```
association :userをFactoryBot内に記述しましょう！

この記述によって商品のインスタンスが生成したと同時に、関連するUserのインスタンスも生成されます。

出品された商品に対しては、必ずUserが紐付いている必要があるため、このように記述する必要があります。（Userは商品を必ず持っているわけではないため、users.rbには記述しません。）

KenyaMatsumoto commented on app/models/item.rb:
```diff
@@ -0,0 +1,32 @@
+  belongs_to :shipping_area
+  belongs_to :shipping_day
+  has_one_attached :image
+
```
userとのアソシエーションを記述しましょう！

KenyaMatsumoto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,112 @@
+        @item.price = ''
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price can't be blank")
+      end
```
アクティブハッシュの「1では登録できない」という内容に関して、数値を代入してテストを行わないといけませんので、シングルクォーテーションをすべて外して、テストの実行お願いします。

KenyaMatsumoto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,112 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price can't be blank")
+      end
+      it 'priceが¥300~¥9,999,999の範囲ないでないと登録できない' do
+        @item.price = 299
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not included in the list')
+      end
```
```ruby
 it 'priceが¥300~¥9,999,999の範囲ないでないと登録できない' do
        @item.price = 299
        @item.valid?
        expect(@item.errors.full_messages).to include('Price is not included in the list')
end
```
こちら厳重なテスト内容にするために、priceが¥10,000,000異常だった場合もテストして頂く必要がございます。
したがって```priceが¥299以下では登録できない```という文言に変更し、``` priceが¥10,000,000以上では登録できない```という内容のテストも追加しましょう！

KenyaMatsumoto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,112 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not included in the list')
+      end
+      it 'priceが半角英数字でないと登録できない' do
```
```  it 'priceが全角英数字でないと登録できない' do　 ```という文言に変更しましょう！

KenyaMatsumoto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,112 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Image can't be blank")
+      end
+    end
```
更に厳重なテスト内容にするために、``` userが紐付いていないと登録できない ```のテストも実行しましょう！

KenyaMatsumoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。

10点コメントいたしましたので、ご確認お願いします。

修正頂いた後、コミット→プッシュ→フォームの再提出をお願いします！
ご不明点や疑問点ございましたら、コネクトメンターにお気軽にご相談ください！
どうぞよろしくおねがいします。

shiki0501 committed:
コードレビュー修正

authenticate_user!, only: :create　追加
paramsパラメータへmerge(user_id)追加
 association にuserを追加
データベースのバリデーション　user,foreign_key: true追加

単体テストコード修正、追加

shiki0501 commented:
コードレビュー、ご指摘ありがとうございました。
致命的なミスであるuserとのアソシエーションを未実装であったことが悔やまれます。。
また、細かな部分までご指摘いただき誠にありがとうございました！
- テスト結果の画像
https://gyazo.com/e4f17c89506ed85f68d4a30bf5aaa3ea

KenyaMatsumoto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,117 @@
+        expect(@item.errors.full_messages).to include('Price is not included in the list')
+      end
+      it 'priceが¥10,000,000以上では登録できない' do
+        @item.price = 10, 0o00, 0o00
```
```ruby
 @item.price = 10, 0o00, 0o00
```
こちら10000000に修正、お願いたします！

KenyaMatsumoto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,117 @@
+        expect(@item.errors.full_messages).to include("Image can't be blank")
+      end
+    end
+  end
```
大変恐れ入りますが、[こちら](https://github.com/shiki0501/furima-35482/pull/3/files#r676104613)の修正分の差分が確認できないかと存じます。
コミット・プッシュ忘れがないかのご確認お願いたします。

KenyaMatsumoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
細かい点になり恐縮ですが、更にコード内容をより良くするために
2点コメントいたしましたので、ご確認お願いします。
実装完了まであと少しになります、もう一踏ん張り頑張っていきましょう！
修正頂いた後、コミット→プッシュ→フォームの再提出をお願いします！
ご不明点や疑問点ございましたら、コネクトメンターにお気軽にご相談ください！
どうぞよろしくおねがいします。

shiki0501 committed:
単体テストコード修正

誤字修正
ユーザーが紐ついて無いと登録できない追加

shiki0501 commented:
大変失礼いたしました。修正事項を見逃しておりました・・。
修正いたしましたので再度コードレビューのほど、よろしくお願いいたします。
- テストコード
https://gyazo.com/559ece9e9ce6a3741cd3211011ea2a43

TakashiKozuma commented:
レビュー依頼ありがとうございます。
全ての修正を確認いたしました。
LGTMです！実装お疲れさまでした。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
